File: backend/app/llm.py
"""LLM クライアント — OpenAI 互換 Chat Completions エンドポイントを使用する。

既定は Groq(無料枠・高速・OpenAI互換)。`LLM_BASE_URL` を差し替えるだけで
Cerebras / OpenRouter / ローカル互換サーバ等へ移行できる(ベンダーロックイン回避)。
返答生成はすべてクラウドのストリーミングに委ねる。`LLM_API_KEY` は必須。
"""

from __future__ import annotations

import json
import os
from collections.abc import AsyncIterator

import httpx

# 既定 = Groq(https://console.groq.com で無料キー取得)。
LLM_BASE_URL = os.environ.get("LLM_BASE_URL", "https://api.groq.com/openai/v1").rstrip("/")
# 日本語雑談向けは Qwen 系が強い。安定して存在する Llama 3.3 70B を既定に。
# 日本語重視なら .env で LLM_MODEL=qwen/qwen3-32b 等に変更可。
LLM_MODEL = os.environ.get("LLM_MODEL", "llama-3.3-70b-versatile")
# 後方互換: LLM_API_KEY が無ければ GROQ_API_KEY を見る。
LLM_API_KEY = os.environ.get("LLM_API_KEY") or os.environ.get("GROQ_API_KEY", "")
TEMPERATURE = 0.85
REQUEST_TIMEOUT = 120


def provider() -> str:
    """状態表示用。使用中のモデル名を返す。"""
    return LLM_MODEL


def _require_api_key() -> str:
    if not LLM_API_KEY:
        raise RuntimeError(
            "LLM_API_KEY が未設定です。backend/.env に LLM のAPIキー(既定はGroq)を設定してください。"
        )
    return LLM_API_KEY


async def stream_chat(
    system: str, messages: list[dict]
) -> AsyncIterator[str]:
    """messages: [{"role": "user"|"assistant", "content": str}, ...] をクラウドへ流す。"""
    api_key = _require_api_key()
    payload_messages = (
        [{"role": "system", "content": system}] if system else []
    ) + messages
    payload = {
        "model": LLM_MODEL,
        "messages": payload_messages,
        "temperature": TEMPERATURE,
        "stream": True,
    }
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }
    url = f"{LLM_BASE_URL}/chat/completions"
    async with httpx.AsyncClient(timeout=REQUEST_TIMEOUT) as client:
        async with client.stream("POST", url, json=payload, headers=headers) as response:
            response.raise_for_status()
            async for line in response.aiter_lines():
                if not line.startswith("data: "):
                    continue
                data = line[len("data: "):].strip()
                if data == "[DONE]":
                    break
                try:
                    chunk = json.loads(data)
                except json.JSONDecodeError:
                    continue
                for choice in chunk.get("choices", []):
                    text = choice.get("delta", {}).get("content")
                    if text:
                        yield text


async def complete(prompt: str) -> str:
    """非ストリーミングの単発補完(事実抽出・日記・要約・声かけ用)。"""
    parts: list[str] = []
    async for chunk in stream_chat("", [{"role": "user", "content": prompt}]):
        parts.append(chunk)
    return "".join(parts).strip()

File: worker/src/llm.ts
/** LLM クライアント — OpenAI 互換 Chat Completions(既定 Groq)。backend/app/llm.py の移植。
 *
 * fetch のストリーミング応答(ReadableStream)を SSE 行としてパースし、delta.content を順次返す。
 * LLM 待ちは I/O であり Workers の CPU 時間(無料10ms)には計上されない。 */

import type { Env } from "./env";

const TEMPERATURE = 0.85;

export interface ChatMessage {
  role: string;
  content: string;
}

function apiKey(env: Env): string {
  const key = env.LLM_API_KEY;
  if (!key) {
    throw new Error("LLM_API_KEY が未設定です。`wrangler secret put LLM_API_KEY` で設定してください。");
  }
  return key;
}

export function provider(env: Env): string {
  return env.LLM_MODEL;
}

/** system + messages をクラウドへ流し、テキスト片を yield する非同期ジェネレータ。
 *
 * maxTokens: 未指定なら無制限。ペルソナ指示(「1〜3文」等)はモデルが無視することがあるため、
 * 会話のテンポを守る必要がある呼び出し元では明示的に上限を渡すこと。 */
export async function* streamChat(
  env: Env,
  system: string,
  messages: ChatMessage[],
  maxTokens?: number,
): AsyncGenerator<string> {
  const payloadMessages = (system ? [{ role: "system", content: system }] : []).concat(messages);
  const body = JSON.stringify({
    model: env.LLM_MODEL,
    messages: payloadMessages,
    temperature: TEMPERATURE,
    stream: true,
    ...(maxTokens !== undefined ? { max_tokens: maxTokens } : {}),
  });
  const url = `${env.LLM_BASE_URL.replace(/\/+$/, "")}/chat/completions`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey(env)}`,
      "Content-Type": "application/json",
    },
    body,
  });
  if (!response.ok || !response.body) {
    const text = await response.text().catch(() => "");
    throw new Error(`LLM 応答エラー (${response.status}): ${text.slice(0, 200)}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith("data:")) continue;
      const data = trimmed.slice(trimmed.indexOf(":") + 1).trim();
      if (data === "[DONE]") return;
      try {
        const chunk = JSON.parse(data) as {
          choices?: { delta?: { content?: string } }[];
        };
        for (const choice of chunk.choices ?? []) {
          const text = choice.delta?.content;
          if (text) yield text;
        }
      } catch {
        // 不完全な JSON 断片は無視(次チャンクで補完される)
      }
    }
  }
}

/** 非ストリーミングの単発補完(事実抽出・日記・要約・声かけ用)。 */
export async function complete(env: Env, prompt: string): Promise<string> {
  const parts: string[] = [];
  for await (const chunk of streamChat(env, "", [{ role: "user", content: prompt }])) {
    parts.push(chunk);
  }
  return parts.join("").trim();
}
